lib/linked_list.rb

```diff
@@ -12,148 +11,355 @@ def initialize(value, next_node = nil)
+  # Time Complexity: O(n)
+  # >> traverse length of LL to reverse all nodes
+  # Space Complexity: O(1)
+  # >> store previous, current, and next nodes
+  def reverse
+    return nil if @head.nil?
+
+    current_node = @head
+    previous_node = nil
+    next_node = nil
+
+    while current_node
+      next_node = current_node.next
+      current_node.next = previous_node
+      previous_node = current_node
+      current_node = next_node
     end
 
-    # method that returns the value of the last node in the linked list
-    # returns nil if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_last
-      raise NotImplementedError
+    @head = previous_node
+  end
+
+  # method that returns the value of the last node in the linked list
+  # returns nil if the linked list is empty
+  # Time Complexity: O(n)
+  # >> traverse length of LL to get to last node
+  # Space Complexity: O(1)
+  # >> store current node
+  def get_last
+    return nil if @head.nil?
+
+    current_node = @head
+
+    until current_node.next.nil?
+      current_node = current_node.next
     end
+
+    return current_node.data
+  end
   
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_middle_value
-      raise NotImplementedError
-    end
-
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_nth_from_end(n)
-      raise NotImplementedError
-    end
-
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def has_cycle
-      raise NotImplementedError
-    end
-
-    # method to insert a new node with specific data value, assuming the linked
-    # list is sorted in ascending order
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def insert_ascending(value)
-      raise NotImplementedError
-    end
-
-    # Helper method for tests
-    # Creates a cycle in the linked list for testing purposes
-    # Assumes the linked list has at least one node
-    def create_cycle
-      return if @head == nil # don't do anything if the linked list is empty
-
-      # navigate to last node
-      current = @head
-      while current.next != nil
-          current = current.next
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(n)
+  # >> get length of LL (pass through all n nodes)
+  # >> go through half of those nodes
+  # Space Complexity: O(1)
+  # >> store mid_index, current_node, current_node_num
+  def find_middle_value
+    return nil if @head.nil?
+
+    mid_index = length / 2
+    current_node = @head
+    current_node_num = 0

```

Comment:
This works, but could you do something similar to the method to find a cycle.

Reply:
Ah I see - I think India mentioned in the previous CS Fun session that you could move a slow and fast pointer and when fast reaches the end, slow will be at the middle